### PR TITLE
Manager for Capture Request Timer

### PR DESCRIPTION
**Pull request by mgacy**
Protocol, implementation, and tests for an object responsible for managing the timer for `CapturedRequestCollector`. The primary motivation for this design is to avoid the ugliness of real timers while testing the request collector.

### CODE REVIEW

**Review by jagreenwood — APPROVED** [review]
Looks good. A couple of nitpicky things

File: Sources/BlueTriangle/NetworkCapture/CaptureTimerManager.swift [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            switch self {
+            case .inactive:
+                return 0
+            case .active(let span):
+                return span
+            }
+        }
+    }
+
+    enum Action {
```
[the remark this answers is not in the document]
Comment (mgacy) [reply Sources/BlueTriangle/NetworkCapture/CaptureTimerManager.swift:25]: Good call

File: Sources/BlueTriangle/NetworkCapture/CaptureTimerManager.swift [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        case active(span: Int)
+
+        var span: Int {
+            switch self {
+            case .inactive:
+                return 0
+            case .active(let span):
+                return span
+            }
+        }
```
[the remark this answers is not in the document]
Comment (mgacy) [reply Sources/BlueTriangle/NetworkCapture/CaptureTimerManager.swift:22]: I guess it's really a remnant of the design phase. The intention was just to provide access to that part of state without the ugliness of pattern matching, but it's not really needed; will remove

File: Sources/BlueTriangle/NetworkCapture/CaptureTimerManager.swift [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                state = .inactive
+            }
+        case (.inactive, .fire):
+            #if DEBUG
+            print("\(#function) - Unanticipated timer fire.")
+            #endif
+            timer?.cancel()
+            timer = nil
+        case (.inactive, .pause):
+            return
```
[the remark this answers is not in the document]
Comment (mgacy) [reply Sources/BlueTriangle/NetworkCapture/CaptureTimerManager.swift:98]: The manager's owner can read its `state` but I don't see the owner as needing to have any knowledge of it. I do expect situations where the owner calls `cancel()` after the manager has already triggered the `.active -> .inactive` transition due to the timer having fired `configuration.spanCount` times. In this situation, we can be reasonably confident the timer has already been cancelled and `timer` set to `nil`.

My treatment of the `(.inactive, .fire)` case was prompted by uncertainty about possible behavior. Given the potential for undefined dispatch object states to cause crashes, I opted to treat the unexpected `(.inactive, .fire)` more defensively and ensure that the timer was cancelled and deallocated.

Your question prompted me to do some more testing and I was able to observe the `(.inactive, .fire)` case when using multiple `DispatchQueue`s:

```swift
func testConcurrentAccess1() throws {
    let configuration = NetworkCaptureConfiguration(
        spanCount: 10,
        initialSpanDuration: 0.00001,
        subsequentSpanDuration: 0.00001)

    let managerQueue = DispatchQueue(label: "com.bluetriangle.uploader",
                                     qos: .userInitiated,
                                     autoreleaseFrequency: .workItem)

    var manager: CaptureTimerManager? = CaptureTimerManager(
        queue: managerQueue,
        configuration: configuration,
        timerLeeway: Self.timerLeeway)
    manager?.handler = { print("----------FIRED----------") }

    let iterationCount = 100_000

    let group = DispatchGroup()
    var n = 0
    DispatchQueue.global().async(group: group) {
        for i in 0 ..< iterationCount {
            if (i % 2 == 0) {
                print("start")
                manager?.start()
            } else {
                if (i % 3 == 0) {
                    print("end")
                    manager?.cancel()
                }
                n += 1
            }
        }
    }
    ...
    
```

Crashes were more likely, however.

**Pull request merged**